### ISSUE TITLE
boolean row-wise reductions?

### ISSUE THREAD

**Issue by MarcoGorelli** [data-apis/dataframe-api#124]
Currently these can be done, but they're a bit clunky

e.g. of how one might do `dropna` using the current api:
```python
columns = df.get_column_names()
mask = ~df.get_column_by_name(columns[0]).isnan()
for col in columns[1:]:
    mask &= ~df.get_column_by_name(col).isnan()
result = df.get_rows_by_mask(mask)
```

It might be good to have something that takes a boolean DataFrame, like
```
df.isnan()
```
and returns a 1-D Column which reduced the DataFrame by row, either with `any` or `all`

OK with not having `DataFrame.mean(axis=1)` as part of the standard, but for boolean values this seems like a fairly important task for filtering?

e.g. of where I came across this:

https://github.com/mwaskom/seaborn/blob/5d9f37159bbd3ac44c8c8a06825583ba25648525/seaborn/relational.py#L520

**Comment by rgommers** [comment]
The line of Seaborn code you pointed at is `y = data.get("y", empty)`. Did you mean `self.plot_data.dropna()` higher up, or something else?

For `.dropna()` I assume that that's only an example, and that is a method that should be added to the API rather than reimplemented in user land?

I'm not 100% clear about the ask here. There are `.any` and `.all` methods: https://github.com/data-apis/dataframe-api/blob/d10a096ccc1612ec4acf8503aff58b3acb4e3738/spec/API_specification/dataframe_api/dataframe_object.py#LL467-L477C12. 

You are thinking that those need to be able to be used row-wise in addition to column-wise? If so, that sounds reasonable to me.

**Comment by MarcoGorelli** [comment]
for me it shows the `dropna` line:

![image](https://user-images.githubusercontent.com/33491632/228909706-a438bac1-1330-4e8f-a898-106746a9a01d.png)

anyway, yes,  it's that one :) 

> There are .any and .all methods

Yes, but they do something different - I'll show an example

If we start with
```python
       a      b      c
0   True  False  False
1  False   True  False
2  False  False  False
```
then `df.any()` would give
```python
      a     b      c
0  True  True  False
```

What I think we need, for filtering purposes, is something which would give a `Column` containing
```
True
True
False
```
so we can pass it to `get_rows_by_mask`

**Comment by rgommers** [comment]
Looks like we'll want separate methods `.any_rowwise`/`all_rowwise` which work only for a dataframe consisting of columns that all have `bool` dtype, returning a `Column`, and supporting a `skip_na` keyword to deal with nulls.

**Comment by kkraus14** [comment]
I vaguely remember an earlier conversation where the decision was made to punt on supporting row-wise operations for the time being given the difficulty in implementing them nicely for many dataframe implementations.

Did we rule out a DataFrame level `dropna` function?

**Comment by rgommers** [comment]
> I vaguely remember an earlier conversation where the decision was made to punt on supporting row-wise operations for the time being given the difficulty in implementing them nicely for many dataframe implementations.

Yes indeed, and that's still the intent, and that's why this proposal is only a very specific any/all method for boolean mask construction only. The case for that seems compelling, and it's being done here in as restricted a form as possible (only any/all, only bool dtype) to avoid encouraging other row-wise reduction APIs.



> Did we rule out a DataFrame level `dropna` function?

No. And that seems useful.

This case isn't just for dropping NAs though, think about things like outlier filtering - say you want a column to use as a mask after identifying outliers in multiple columns with a thresholding operation a la `values > 999`.
